src/dbnode/storage/mediator.go

```diff
@@ -182,54 +203,83 @@ func (m *mediator) Close() error {
 // is potentially still on disk (if it hasn't been cleaned up yet).
 //
 // See comment over mediatorTimeBarrier for more details on how this is implemented.
-func (m *mediator) ongoingFilesystemProcesses() {
-	log := m.opts.InstrumentOptions().Logger()
+func (m *mediator) ongoingFileSystemProcesses() {
 	for {
 		select {
 		case <-m.closedCh:
 			return
 		default:
 			m.sleepFn(tickCheckInterval)
-			// See comment over mediatorTimeBarrier for an explanation of this logic.
-			mediatorTime, err := m.mediatorTimeBarrier.fsProcessesWait()
-			if err != nil {
-				log.Error("error within ongoingFilesystemProcesses waiting for next mediatorTime", zap.Error(err))
-				continue
+
+			// Check if the db is already closed.
+			if m.databaseState() == databaseClosed {
+				return
 			}
 
-			m.databaseFileSystemManager.Run(mediatorTime, syncRun, noForce)
+			m.runFileSystemProcesses()
 		}
 	}
 }
 
 func (m *mediator) ongoingTick() {
-	var (
-		log          = m.opts.InstrumentOptions().Logger()
-		mediatorTime = m.mediatorTimeBarrier.initialMediatorTime()
-	)
+	mediatorTime := m.mediatorTimeBarrier.initialMediatorTime()
 	for {
 		select {
 		case <-m.closedCh:
 			return
 		default:
 			m.sleepFn(tickCheckInterval)
 
+			// Check if the db is already closed.
+			if m.databaseState() == databaseClosed {
+				return
+			}
+
 			// See comment over mediatorTimeBarrier for an explanation of this logic.
+			log := m.opts.InstrumentOptions().Logger()
 			newMediatorTime, err := m.mediatorTimeBarrier.maybeRelease()
 			if err != nil {
-				log.Error(
-					"ongoing tick was unable to release time barrier", zap.Error(err))
-				continue
+				log.Error("ongoing tick was unable to release time barrier", zap.Error(err))
+				return
 			}
 			mediatorTime = newMediatorTime
 
-			if err := m.Tick(force, mediatorTime); err != nil {
+			// NB(bodu): We may still hit a db closed error here since the db does not wait upon
+			// completion of ticks.
+			if err := m.Tick(force, mediatorTime); err != nil && err != errDatabaseIsClosed {
 				log.Error("error within tick", zap.Error(err))
 			}
 		}
 	}
 }
 
+func (m *mediator) runFileSystemProcesses() {
+	m.Lock()
+	m.fileSystemProcessesState = fileSystemProcessesBusy
+	m.Unlock()
+	defer func() {
+		m.Lock()
+		m.fileSystemProcessesState = fileSystemProcessesIdle
+		m.Unlock()
+	}()
+
+	// See comment over mediatorTimeBarrier for an explanation of this logic.
+	log := m.opts.InstrumentOptions().Logger()
+	mediatorTime, err := m.mediatorTimeBarrier.fsProcessesWait()
+	if err != nil {
+		log.Error("error within ongoingFileSystemProcesses waiting for next mediatorTime", zap.Error(err))
+		return
+	}
+
+	m.databaseFileSystemManager.Run(mediatorTime, syncRun, noForce)
+}
+
+func (m *mediator) databaseState() databaseState {
+	db := m.database.(*db)
```

Comment:
Hm this seems racy without acquiring a lock. 

Could you add a public method to `type Database interface` just called `IsOpen() bool`?

That way we can do the read lock from there too.